# Derivatives 
---
The derivative of a function describes how the output of a function changes when there is a small change in an input variable

The cost  𝐽
  is the output and  𝑤
  is the input variable.


$$
\begin{equation}
\text{if } w \uparrow \epsilon \text{ causes }J(w) \uparrow \text{by }k \times \epsilon \text{ then}  \\
\frac{\partial J(w)}{\partial w} = k \tag{1}
\end{equation}
$$
This just says if you change the input to the function $J(w)$ by a little bit and the output changes by $k$ times that little bit, then the derivative of $J(w)$ is equal to $k$.


$\frac{\partial J(w)}{\partial w} = 2 w$, the derivative of $J(w) = w^2$ 

In [15]:
J = (3)**2
J_epsilon = (3 + 0.001)**2
k = (J_epsilon - J)/0.001    # difference divided by epsilon
print(f"J = {J}, J_epsilon = {J_epsilon}, dJ_dw ~= k = {k} ")

print(f"(dj_dw * 0.001) = {6*0.001} this means that the J will increse by {6*0.001}")

J = 9, J_epsilon = 9.006001, dJ_dw ~= k = 6.000999999999479 
(dj_dw * 0.001) = 0.006 this means that the J will increse by 0.006


The input value $w$ increased by 0.001 , causing the output change from 9 to 9.006001
this says that $k=6$, so $\frac{\partial J(w)}{\partial w} \approx 6$.
